File: 0230.Kth-Smallest-Element-in-a-BST/step2.cpp
// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
  public:
    int kthSmallest(TreeNode *root, int k) {
        int k_th_smallest = -1;
        int num_seen = 0;
        auto traverse = [k, &k_th_smallest, &num_seen](auto &self, TreeNode *node) -> void {
            if (!node) {
                return;
            }
            self(self, node->left);
            if (k_th_smallest >= 0) {
                return;
            }
            num_seen++;
            if (num_seen== k){
                k_th_smallest = node->val;
                return;
            }
            self(self, node->right);
        };

        traverse(traverse, root);
        return k_th_smallest;
    }
};

File: 0230.Kth-Smallest-Element-in-a-BST/step2.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def kthSmallest(self, root: TreeNode | None, k: int) -> int:
        if root is None:
            raise ValueError("root is None")

        num_seen = 0
        stack = [(root, False)]

        while stack:
            node, seen_left = stack.pop()
            if node is None:
                continue
            if seen_left:
                num_seen += 1
                if num_seen == k:
                    return node.val
                stack.append((node.right, False))
                continue
            stack.append((node, True))
            stack.append((node.left, False))

        return None
